File: rust/app/wx/wx_refund_notice/wx_refund_notice_service.rs
use color_eyre::eyre::Result;
use tracing::{info, error};

use chrono::NaiveDate;

use generated::common::context::{
  Options,
  get_req_id,
};

use smol_str::SmolStr;

use generated::common::wx_pay::decode::{
  decode_wx_refund,
  WxRefundNotify,
  WxRefundResource,
};

use rust_decimal::Decimal;

// wx_pay
use generated::wx::wx_pay::wx_pay_dao::find_one_wx_pay;
use generated::wx::wx_pay::wx_pay_model::WxPaySearch;

// wx_refund_notice (generated dao)
use generated::wx::wx_refund_notice::wx_refund_notice_dao::create_wx_refund_notice;
use generated::wx::wx_refund_notice::wx_refund_notice_service::update_tenant_by_id_wx_refund_notice;
use generated::wx::wx_refund_notice::wx_refund_notice_model::{
  WxRefundNoticeInput,
  WxRefundNoticeRefundStatus,
};

// wx_refund (generated dao)
use generated::wx::wx_refund::wx_refund_dao::{
  find_one_wx_refund,
  update_by_id_wx_refund,
};
use generated::wx::wx_refund::wx_refund_model::{
  WxRefundInput,
  WxRefundStatus,
};

static REFUND_NOTIFY_URL: &str = "/api/wx_pay/wx_refund_notice";

/// 微信退款通知
pub async fn wx_refund_notify(
  wx_refund_notify: WxRefundNotify,
  options: Option<Options>,
) -> Result<()> {
  
  info!(
    "{req_id} wx_refund_notify: {wx_refund_notify:?}",
    req_id = get_req_id(),
  );

  let wx_pay_model = find_one_wx_pay(
    Some(WxPaySearch {
      refund_notify_url: Some(REFUND_NOTIFY_URL.into()),
      ..Default::default()
    }),
    None,
    options,
  ).await?;
  
  let wx_pay_model = match wx_pay_model {
    Some(model) => model,
    None => {
      error!(
        "{req_id} No matching wx_pay_model found for refund_notify_url: {REFUND_NOTIFY_URL}",
        req_id = get_req_id(),
      );
      return Ok(());
    }
  };

  let v3_key = wx_pay_model.v3_key;
  
  if v3_key.is_empty() {
    error!(
      "{req_id} v3_key is empty for wx_pay_model with refund_notify_url: {REFUND_NOTIFY_URL}",
      req_id = get_req_id(),
    );
    return Ok(());
  }

  let wx_refund_resource: WxRefundResource = decode_wx_refund(
    &v3_key,
    wx_refund_notify,
  )?;

  info!(
    "{req_id} wx_refund_resource: {wx_refund_resource:?}",
    req_id = get_req_id()
  );

  // 提取字段
  let refund_id = SmolStr::new(&wx_refund_resource.refund_id);
  let out_refund_no = SmolStr::new(&wx_refund_resource.out_refund_no);
  let out_trade_no = SmolStr::new(&wx_refund_resource.out_trade_no);
  let transaction_id = SmolStr::new(&wx_refund_resource.transaction_id);
  let user_received_account = SmolStr::new(&wx_refund_resource.user_received_account);

  let success_time = wx_refund_resource.success_time
    .map(|s| chrono::NaiveDateTime::parse_from_str(&s, "%Y-%m-%dT%H:%M:%S%z"))
    .transpose()?;
  
  if wx_refund_resource.amount.refund > u32::MAX as u64 ||
    wx_refund_resource.amount.payer_refund > u32::MAX as u64
  {
    error!(
      "{req_id} Refund amounts exceed u32::MAX: refund={refund}, payer_refund={payer_refund}",
      req_id = get_req_id(),
      refund = wx_refund_resource.amount.refund,
      payer_refund = wx_refund_resource.amount.payer_refund,
    );
    return Ok(());
  }
  
  // amounts: already 分
  let amount_refund = wx_refund_resource.amount.refund as u32;
  let amount_payer_refund = wx_refund_resource.amount.payer_refund as u32;
  
  // refund_status (enum in resource)
  let refund_status = match wx_refund_resource.refund_status {
    generated::common::wx_pay::RefundStatus::SUCCESS => WxRefundNoticeRefundStatus::Success,
    generated::common::wx_pay::RefundStatus::CLOSED => WxRefundNoticeRefundStatus::Closed,
    generated::common::wx_pay::RefundStatus::PROCESSING => WxRefundNoticeRefundStatus::Processing,
    _ => WxRefundNoticeRefundStatus::Abnormal,
  };

  // 根据 transaction_id 查询 wx_refund，更新其状态
  let wx_refund_model = find_one_wx_refund(
    Some(generated::wx::wx_refund::wx_refund_model::WxRefundSearch {
      transaction_id: Some(transaction_id.clone()),
      ..Default::default()
    }),
    None,
    options,
  ).await?;
  
  let wx_refund_model = match wx_refund_model {
    Some(model) => model,
    None => {
      error!(
        "{req_id} No matching wx_refund_model found for transaction_id: {transaction_id}",
        req_id = get_req_id(),
        transaction_id = transaction_id,
      );
      return Ok(());
    }
  };
  
  let appid = wx_refund_model.appid;
  let mchid = wx_refund_model.mchid;
  let tenant_id = wx_refund_model.tenant_id;
  let attach2 = wx_refund_model.attach2;
  
  let wx_refund_notice_id = create_wx_refund_notice(
    WxRefundNoticeInput {
      appid: Some(appid),
      mchid: Some(mchid),
      tenant_id: Some(tenant_id),
      refund_id: Some(refund_id.clone()),
      out_trade_no: Some(out_trade_no.clone()),
      transaction_id: Some(transaction_id),
      out_refund_no: Some(out_refund_no),
      refund_status: Some(refund_status),
      success_time,
      user_received_account: Some(user_received_account),
      amount_refund: Some(amount_refund),
      amount_payer_refund: Some(amount_payer_refund),
      create_time: Some(chrono::Local::now().naive_local()),
      ..Default::default()
    },
    options,
  ).await?;
  
  update_tenant_by_id_wx_refund_notice(
    wx_refund_notice_id,
    tenant_id,
    options,
  ).await?;
  
  // 更新 wx_refund 的状态
  update_by_id_wx_refund(
    wx_refund_model.id,
    WxRefundInput {
      refund_id: Some(refund_id),
      out_trade_no: Some(out_trade_no),
      status: Some(match refund_status {
        WxRefundNoticeRefundStatus::Success => WxRefundStatus::Success,
        WxRefundNoticeRefundStatus::Closed => WxRefundStatus::Closed,
        WxRefundNoticeRefundStatus::Processing => WxRefundStatus::Processing,
        WxRefundNoticeRefundStatus::Abnormal => WxRefundStatus::Abnormal,
        _ => WxRefundStatus::NoRefund,
      }),
      success_time,
      ..Default::default()
    },
    options,
  ).await?;
  
  if refund_status != WxRefundNoticeRefundStatus::Success {
    error!(
      "{req_id} Refund status is not success: {refund_status:?}",
      req_id = get_req_id(),
      refund_status = refund_status,
    );
    return Ok(());
  }
  
  if attach2.is_empty() {
    error!(
      "{req_id} pay_notice_attach2: attach2 is empty",
      req_id = get_req_id(),
    );
    return Ok(());
  }
  
  let attach2_obj: serde_json::Value = serde_json::from_str(&attach2)?;
  let action = attach2_obj["action"].as_str().unwrap_or_default();
  let payload = &attach2_obj["payload"];
  
  let amt = Decimal::from(amount_refund) / Decimal::from(100);
  
  info!(
    "{req_id} refund_notice_attach2: amt: {amt} {attach2:?}",
    req_id = get_req_id(),
    attach2 = attach2_obj,
  );
  
  if action == "Test" {
    info!(
      "{req_id} refund notice action: {action} payload: {payload:?} wx_refund_notice_id: {wx_refund_notice_id}",
      req_id = get_req_id(),
    );
    return Ok(());
  }
  
  Ok(())
}

File: deno/src/wx/wx_refund_notice/wx_refund_notice.service.ts
import {
  error,
  log,
} from "/lib/context.ts";

import {
  getWxPayModel,
} from "/src/wx/wx_pay/wx_pay.dao.ts";

import {
  findOneWxPay,
} from "/gen/wx/wx_pay/wx_pay.dao.ts";

import {
  createWxRefundNotice,
  updateTenantByIdWxRefundNotice,
} from "/gen/wx/wx_refund_notice/wx_refund_notice.dao.ts";

import {
  findOneWxRefund,
  updateByIdWxRefund,
} from "/gen/wx/wx_refund/wx_refund.dao.ts";

import {
  WxRefundNoticeRefundStatus,
  WxRefundStatus,
} from "/gen/types.ts";

import {
  Decimal,
} from "decimal.js";

import WxPay from "wechatpay-node-v3";

import {
  Buffer,
} from "node:buffer";

import dayjs from "dayjs";

interface Ipay {
  appid: string;
  mchid: string;
  publicKey: string;
  privateKey: string;
  key?: string;
}

interface RefundNoticeResource {
  mchid: string;
  out_trade_no: string;
  transaction_id: string;
  out_refund_no: string;
  refund_id: string;
  refund_status: string;
  success_time?: string;
  user_received_account: string;
  amount: {
    total: number;
    refund: number;
    payer_total: number;
    payer_refund: number;
  };
}

async function fetchRefund_notice(
  obj: Ipay,
  params: {
    // deno-lint-ignore no-explicit-any
    body: any;
    signature: string;
    serial: string;
    nonce: string;
    timestamp: string;
  },
) {
  const wxPay = new WxPay({
    appid: obj.appid,
    mchid: obj.mchid,
    publicKey: Buffer.from(obj.publicKey),
    privateKey: Buffer.from(obj.privateKey),
    key: obj.key,
  });
  const ret = await wxPay.verifySign(params);
  if (!ret) {
    throw new Error("签名验证失败");
  }
  const noticeBody = params.body;
  // 数据密文
  const ciphertext = noticeBody.resource.ciphertext;
  // 附加数据
  const associated_data = noticeBody.resource.associated_data;
  // 随机串
  const nonce = noticeBody.resource.nonce;
  // APIv3密钥
  const key = obj.key;
  const data = wxPay.decipher_gcm<RefundNoticeResource>(
    ciphertext,
    associated_data,
    nonce,
    key,
  );
  return data;
}

/**
 * 退款通知
 */
export async function refund_notice(
  notify_url: string,
  params: {
    // deno-lint-ignore no-explicit-any
    body: any;
    signature: string;
    serial: string;
    nonce: string;
    timestamp: string;
  },
) {
  log(`refund_notice_params: ${ JSON.stringify(params) }`);
  log(`refund_notice_body: ${ JSON.stringify(params.body) }`);
  
  const wx_pay_model0 = await findOneWxPay({
    refund_notify_url: notify_url,
  });
  if (!wx_pay_model0 || !wx_pay_model0.appid) {
    error(`退款回调地址: ${ notify_url } 未配置微信支付`);
    return;
  }
  const wx_pay_model = await getWxPayModel(wx_pay_model0);
  const appid = wx_pay_model.appid;
  const mchid = wx_pay_model.mchid;
  const public_key = wx_pay_model.public_key;
  const private_key = wx_pay_model.private_key;
  const v3_key = wx_pay_model.v3_key;
  
  const result = await fetchRefund_notice(
    {
      appid,
      mchid,
      publicKey: public_key,
      privateKey: private_key,
      key: v3_key,
    },
    params,
  );
  
  log(`refund_notice_result: ${ JSON.stringify(result) }`);
  
  const refund_id = result.refund_id;
  const out_refund_no = result.out_refund_no;
  const out_trade_no = result.out_trade_no;
  const transaction_id = result.transaction_id;
  const user_received_account = result.user_received_account;
  
  let success_time: string | undefined = undefined;
  if (result.success_time) {
    success_time = dayjs(result.success_time).format("YYYY-MM-DD HH:mm:ss");
  }
  
  const amount_total = result.amount.total;
  const amount_refund = result.amount.refund;
  const amount_payer_total = result.amount.payer_total;
  const amount_payer_refund = result.amount.payer_refund;
  
  const refund_status = result.refund_status as unknown as WxRefundNoticeRefundStatus;
  
  // 根据 transaction_id 查询 wx_refund, 更新其状态
  const wx_refund_model = await findOneWxRefund({
    transaction_id,
  });
  
  if (!wx_refund_model) {
    error(`退款通知: 未找到 transaction_id=${ transaction_id } 的退款记录`);
    return;
  }
  
  const tenant_id = wx_refund_model.tenant_id;
  const attach2 = wx_refund_model.attach2;
  
  // 创建退款通知记录
  const wx_refund_notice_id = await createWxRefundNotice({
    appid: wx_refund_model.appid,
    mchid: wx_refund_model.mchid,
    refund_id,
    out_trade_no,
    transaction_id,
    out_refund_no,
    refund_status,
    success_time,
    user_received_account,
    amount_total,
    amount_refund,
    amount_payer_total,
    amount_payer_refund,
  });
  
  await updateTenantByIdWxRefundNotice(
    wx_refund_notice_id,
    tenant_id,
  );
  
  // 更新 wx_refund 的状态
  const wx_refund_status: WxRefundStatus = (() => {
    switch (refund_status) {
      case WxRefundNoticeRefundStatus.Success:
        return WxRefundStatus.Success;
      case WxRefundNoticeRefundStatus.Closed:
        return WxRefundStatus.Closed;
      case WxRefundNoticeRefundStatus.Processing:
        return WxRefundStatus.Processing;
      case WxRefundNoticeRefundStatus.Abnormal:
        return WxRefundStatus.Abnormal;
      default:
        return WxRefundStatus.NoRefund;
    }
  })();
  
  await updateByIdWxRefund(
    wx_refund_model.id,
    {
      refund_id,
      out_trade_no,
      status: wx_refund_status,
      success_time,
    },
  );
  
  if (refund_status !== WxRefundNoticeRefundStatus.Success) {
    error(`退款通知: 退款状态非成功: ${ refund_status }`);
    return;
  }
  
  if (!attach2) {
    error(`refund_notice_attach2: attach2 为空!`);
    return;
  }
  
  const attach2Obj = JSON.parse(attach2);
  const action = attach2Obj.action;
  
  const amt = new Decimal(amount_payer_refund).div(100);
  
  log(`refund_notice_attach2: action: ${ action } amt: ${ amt } ${ JSON.stringify(attach2Obj) }`);
  
  // TODO: handle action
}
